test/test.js

```diff
@@ -52,110 +53,118 @@ const extractHashes = (assets, filter) => Object.keys(assets)
-                expect(commons.source()).to.contain(hash);
-            });
-        }));
-
-    it('Works with manifest file', () => webpackCompile('manifest')
-        .then((stats) => {
-            expectAssetsNameToContainHash(stats);
-
-            const hashes = extractHashes(stats.compilation.assets, n => n[0] !== 'manifest');
-            const commons = findAssetByName(stats.compilation.assets, 'manifest');
-
-            expect(hashes).to.have.lengthOf(2);
-            hashes.forEach((hash) => {
-                expect(commons.source()).to.contain(hash);
-            });
-        }));
-
-    it('Works with a complex uglified project', () => webpackCompile('manifest-uglify')
-        .then((stats) => {
-            expectAssetsNameToContainHash(stats);
-
-            const hashes = extractHashes(stats.compilation.assets, n => n[0] !== 'manifest');
-            const commons = findAssetByName(stats.compilation.assets, 'manifest');
-
-            expect(hashes).to.have.lengthOf(2);
-            hashes.forEach((hash) => {
-                expect(commons.source()).to.contain(hash);
-            });
-        }));
-
-    it('Works with HTML output', () => webpackCompile('html')
-        .then((stats) => {
-            expectAssetsNameToContainHash(stats, name => name !== 'index.html');
-
-            const hashes = extractHashes(stats.compilation.assets, n => n[0] !== 'index');
-            const commons = findAssetByName(stats.compilation.assets, 'index');
-
-            expect(hashes).to.have.lengthOf(2);
-            hashes.forEach((hash) => {
-                expect(commons.source()).to.contain(hash);
-            });
-        }));
-
-    it('Works with sourcemaps', () => webpackCompile('sourcemap')
-        .then((stats) => {
-            const assets = stats.compilation.assets;
-
-            // New hash
-            expect(assets).to.have.property('entry.d36dd5d3312b77a32d66.js');
-
-            // Sourcemap still uses the old hash
-            expect(assets).to.have.property('entry.c72f41ea29f35ab86a6b.js.map');
-
-            // Source code still points to the old sourcemap
-            expect(assets['entry.d36dd5d3312b77a32d66.js'].source())
-                .to.contain('sourceMappingURL=entry.c72f41ea29f35ab86a6b.js.map');
-
-            // But sourcemaps has the name of the new source code file
-            expect(assets['entry.c72f41ea29f35ab86a6b.js.map'].source())
-                .to.contain('entry.d36dd5d3312b77a32d66.js');
-        }));
-
-    it('Works with code splitting', () => webpackCompile('code-split')
-        .then((stats) => {
-            const main = findAssetByName(stats.compilation.assets, 'main');
-            const onDemandChunk = stats.compilation.chunks.filter(c => c.name === null)[0];
-
-            // Source code uses new hash for on-demand chunk
-            expect(main.source())
-                .to.contain(onDemandChunk.renderedHash);
-        }));
+    const modes = ['development', 'production'];
```

Comment:
Why do we need two modes?

Reply:
webpack 4 hides a lot of configuration behind the modes now, the idea is to reduce the configuration for the user.

The main change between development and production is that production provides minification. There is also apparently a different in the way the chunk graph is generated as the order of the chunks were different between development and production.

Reply:
Got it, thanks for explaining